File: connector/src/testflying_connector/rate_limit.py
from __future__ import annotations

import math
import re
import time
from collections import deque
from dataclasses import dataclass
from threading import Lock

from testflying_connector.config import Settings


@dataclass(frozen=True)
class RateLimitRule:
    key: str
    max_requests: int
    window_seconds: int


@dataclass(frozen=True)
class RateLimitDecision:
    allowed: bool
    retry_after: int
    rule: RateLimitRule


class SlidingWindowRateLimiter:
    def __init__(self) -> None:
        self._events: dict[str, deque[float]] = {}
        self._lock = Lock()

    def check(self, rule: RateLimitRule) -> RateLimitDecision:
        now = time.monotonic()
        max_requests = max(rule.max_requests, 1)
        window_seconds = max(rule.window_seconds, 1)
        with self._lock:
            events = self._events.setdefault(rule.key, deque())
            cutoff = now - window_seconds
            while events and events[0] <= cutoff:
                events.popleft()
            if len(events) >= max_requests:
                retry_after = max(math.ceil(events[0] + window_seconds - now), 1)
                return RateLimitDecision(
                    allowed=False,
                    retry_after=retry_after,
                    rule=RateLimitRule(rule.key, max_requests, window_seconds),
                )
            events.append(now)
        return RateLimitDecision(
            allowed=True,
            retry_after=0,
            rule=RateLimitRule(rule.key, max_requests, window_seconds),
        )

    def reset(self) -> None:
        with self._lock:
            self._events.clear()


class StoreRateLimitPolicy:
    def __init__(self, settings: Settings) -> None:
        self._settings = settings
        self._apple_max_requests = max(settings.apple_rate_limit_fallback_max_requests, 1)
        self._apple_window_seconds = max(settings.apple_rate_limit_window_seconds, 1)
        self._lock = Lock()

    def rule_for_platform(self, platform: str) -> RateLimitRule:
        normalized_platform = platform.strip().lower()
        if normalized_platform == "ios":
            with self._lock:
                return RateLimitRule(
                    key="apple",
                    max_requests=self._apple_max_requests,
                    window_seconds=self._apple_window_seconds,
                )
        if normalized_platform == "android":
            return RateLimitRule(
                key="google",
                max_requests=self._settings.google_rate_limit_max_requests,
                window_seconds=self._settings.google_rate_limit_window_seconds,
            )
        return RateLimitRule(
            key="unknown-store",
            max_requests=min(
                self._settings.google_rate_limit_max_requests,
                self._apple_max_requests,
            ),
            window_seconds=max(
                self._settings.google_rate_limit_window_seconds,
                self._apple_window_seconds,
            ),
        )

    def record_apple_rate_limit_header(self, header_value: str | None) -> None:
        user_hour_limit = parse_apple_user_hour_limit(header_value)
        if user_hour_limit is None:
            return
        adjusted_limit = max(
            math.floor(user_hour_limit * self._settings.apple_rate_limit_safety_ratio),
            1,
        )
        with self._lock:
            self._apple_max_requests = adjusted_limit
            self._apple_window_seconds = 3600


def parse_apple_user_hour_limit(header_value: str | None) -> int | None:
    if header_value is None:
        return None
    match = re.search(r"(?:^|;)\s*user-hour-lim\s*:\s*(\d+)\s*(?:;|$)", header_value)
    if match is None:
        return None
    return int(match.group(1))

File: connector/internal/connector/ratelimit.go
package connector

import (
	"math"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

type RateLimitRule struct {
	Key         string
	MaxRequests int
	Window      time.Duration
}

type RateLimitDecision struct {
	Allowed    bool
	RetryAfter time.Duration
	Rule       RateLimitRule
}

type SlidingWindowRateLimiter struct {
	mu     sync.Mutex
	events map[string][]time.Time
	now    func() time.Time
}

func NewSlidingWindowRateLimiter() *SlidingWindowRateLimiter {
	return &SlidingWindowRateLimiter{
		events: make(map[string][]time.Time),
		now:    time.Now,
	}
}

func (l *SlidingWindowRateLimiter) Check(rule RateLimitRule) RateLimitDecision {
	maxRequests := rule.MaxRequests
	if maxRequests < 1 {
		maxRequests = 1
	}
	window := rule.Window
	if window <= 0 {
		window = time.Second
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	now := l.now()
	cutoff := now.Add(-window)
	events := l.events[rule.Key]
	kept := events[:0]
	for _, event := range events {
		if event.After(cutoff) {
			kept = append(kept, event)
		}
	}
	events = kept

	if len(events) >= maxRequests {
		retryAfter := events[0].Add(window).Sub(now)
		if retryAfter < time.Second {
			retryAfter = time.Second
		}
		l.events[rule.Key] = events
		return RateLimitDecision{
			Allowed:    false,
			RetryAfter: retryAfter,
			Rule: RateLimitRule{
				Key:         rule.Key,
				MaxRequests: maxRequests,
				Window:      window,
			},
		}
	}

	events = append(events, now)
	l.events[rule.Key] = events
	return RateLimitDecision{
		Allowed: true,
		Rule: RateLimitRule{
			Key:         rule.Key,
			MaxRequests: maxRequests,
			Window:      window,
		},
	}
}

func (l *SlidingWindowRateLimiter) Reset() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.events = make(map[string][]time.Time)
}

type StoreRateLimitPolicy struct {
	settings         Settings
	mu               sync.Mutex
	appleMaxRequests int
	appleWindow      time.Duration
}

func NewStoreRateLimitPolicy(settings Settings) *StoreRateLimitPolicy {
	return &StoreRateLimitPolicy{
		settings:         settings,
		appleMaxRequests: maxInt(settings.AppleRateLimitFallbackMax, 1),
		appleWindow:      maxDuration(settings.AppleRateLimitWindow, time.Second),
	}
}

func (p *StoreRateLimitPolicy) RuleForPlatform(platform string) RateLimitRule {
	switch normalizePlatform(platform) {
	case "ios":
		p.mu.Lock()
		defer p.mu.Unlock()
		return RateLimitRule{
			Key:         "apple",
			MaxRequests: p.appleMaxRequests,
			Window:      p.appleWindow,
		}
	case "android":
		return RateLimitRule{
			Key:         "google",
			MaxRequests: p.settings.GoogleRateLimitMaxRequests,
			Window:      p.settings.GoogleRateLimitWindow,
		}
	default:
		return RateLimitRule{
			Key:         "unknown-store",
			MaxRequests: minInt(p.settings.GoogleRateLimitMaxRequests, p.appleMaxRequests),
			Window:      maxDuration(p.settings.GoogleRateLimitWindow, p.appleWindow),
		}
	}
}

func (p *StoreRateLimitPolicy) RecordAppleRateLimitHeader(headerValue string) {
	limit, ok := ParseAppleUserHourLimit(headerValue)
	if !ok {
		return
	}
	adjusted := int(math.Floor(float64(limit) * p.settings.AppleRateLimitSafetyRatio))
	if adjusted < 1 {
		adjusted = 1
	}
	p.mu.Lock()
	defer p.mu.Unlock()
	p.appleMaxRequests = adjusted
	p.appleWindow = time.Hour
}

var appleUserHourLimitPattern = regexp.MustCompile(`(?:^|;)\s*user-hour-lim\s*:\s*(\d+)\s*(?:;|$)`)

func ParseAppleUserHourLimit(headerValue string) (int, bool) {
	match := appleUserHourLimitPattern.FindStringSubmatch(headerValue)
	if len(match) != 2 {
		return 0, false
	}
	value, err := strconv.Atoi(match[1])
	if err != nil {
		return 0, false
	}
	return value, true
}

func normalizePlatform(platform string) string {
	switch strings.ToLower(strings.TrimSpace(platform)) {
	case "ios":
		return "ios"
	case "android":
		return "android"
	default:
		return ""
	}
}

func minInt(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxDuration(a time.Duration, b time.Duration) time.Duration {
	if a > b {
		return a
	}
	return b
}
